relay-server/src/actors/processor.rs

```diff
@@ -951,77 +962,333 @@ impl EnvelopeProcessorService {
+        });
+
+        event::finalize(state, &self.inner.config)?;
+        self.light_normalize_event(state)?;
+        dynamic_sampling::normalize(state);
+        event::filter(state)?;
+        dynamic_sampling::run(state, &self.inner.config);
+
+        // We avoid extracting metrics if we are not sampling the event while in non-processing
+        // relays, in order to synchronize rate limits on indexed and processed transactions.
+        if self.inner.config.processing_enabled() || state.sampling_result.should_drop() {
+            self.extract_metrics(state)?;
+        }
+
+        dynamic_sampling::sample_envelope(state)?;
 
-            if_processing!({
-                unreal::expand(state, &self.inner.config)?;
+        if_processing!(self.inner.config, {
+            event::store(state, &self.inner.config, self.inner.geoip_lookup.as_ref())?;
+            self.enforce_quotas(state)?;
+            profile::process(state, &self.inner.config);
+        });
+
+        if state.has_event() {
+            event::scrub(state)?;
+            event::serialize(state)?;
+            if_processing!(self.inner.config, {
+                span::extract_from_event(state);
             });
+        }
 
+        attachment::scrub(state);
+        Ok(())
+    }
+    fn process_attachments(&self, state: &mut ProcessEnvelopeState) -> Result<(), ProcessingError> {
+        // Only some attachments types have create event set to true.
+        if state.creates_event() {
             event::extract(state, &self.inner.config)?;
-            profile::transfer_id(state);
 
-            if_processing!({
-                unreal::process(state)?;
+            if_processing!(self.inner.config, {
                 attachment::create_placeholders(state);
             });
 
             event::finalize(state, &self.inner.config)?;
             self.light_normalize_event(state)?;
-            dynamic_sampling::normalize(state);
             event::filter(state)?;
-            dynamic_sampling::run(state, &self.inner.config);
-
-            // We avoid extracting metrics if we are not sampling the event while in non-processing
-            // relays, in order to synchronize rate limits on indexed and processed transactions.
-            if self.inner.config.processing_enabled() || state.sampling_result.should_drop() {
-                self.extract_metrics(state)?;
-            }
 
-            dynamic_sampling::sample_envelope(state)?;
-
-            if_processing!({
+            if_processing!(self.inner.config, {
                 event::store(state, &self.inner.config, self.inner.geoip_lookup.as_ref())?;
             });
         }
 
-        if_processing!({
+        if_processing!(self.inner.config, {
             self.enforce_quotas(state)?;
-            profile::process(state, &self.inner.config);
-            self.process_check_ins(state);
-            span::process(state, self.inner.config.clone());
         });
 
         if state.has_event() {
```

Comment:
Can you double-check the conditions for this? There should be an invariant that `state.has_event() === state.creates_event()` earlier. To be precise, when the state creates an event, there should be a placeholder created, which must be serialized unconditionally. If this is not the case, we likely have an implementation bug in `creates_event()`.

Reply:
Currently `enforce_quotas` can remove the event from the state, that's why this condition still can be `false` or `true`. We can address this in  the followup.